docs/concepts/core-workflow-concepts.md

```diff
@@ -99,7 +101,7 @@ A condition is a test that's evaluated to determine whether a workflow instance
 
 When evaluated, conditions are provided with the workflow instance and the trigger that has caused the evaluation. It is a simple true/false test, and the workflow engine contains the following pre-defined conditions:
 - TriggerContentTypeCondition - validates that the content type of the incoming trigger exactly matches a given content type.
-- InvokeExternalServiceCondition - performs a GET or POST request (the method is configurable) to an external endpoint and evaluates the return value, which is expected to be a string containing either "true" or "false".
+- InvokeExternalServiceCondition - performs a GET or POST request (the method is configurable) to an external endpoint and evaluates the return value, which is expected to be a string containing either "true" or "false". Anything other than a 200 status code results in an exception being thrown.
```

Comment:
Even a 202? (Which is what you'll typically get from an endpoint that kicks off an Azure Functions orchestration.)

Reply:
Well, there's a thing.

The code in `InvokeExternalServiceCondition` uses `HttpResponseMessage.IsSuccessStatusCode`, which returns true for all 2XX status codes. So what I've written is wrong for the code as it stands.

However... I would argue that the code is wrong. Our code does not currently support long running ops, when invoking external service conditions or actions, and whilst I can see that you might want the workflow engine to invoke an action that kicks something off, and not necessarily care whether it's long running or not as long as it returns a 2XX code, I don't think the same should apply for a condition.

Conditions should evaluate to "true" or "false", and unless we explicitly add support for long running ops in condition evaluation, I don't think we should be accepting anything other than a 200. 

Thoughts?

Reply:
Ah, I'd somehow blanked the fact that this was for a condition.

I think we should define what happens if the response does not have a body containing either `true` or `false`: that should put the workflow into a faulted state, I think. And I think by extension you're right, in this particular case we also need to handle certain 'success' outcomes as wrong, and therefore causing a faulted state. I think a 202 should be one of those, because as the HTTP spec says, when the response is a 202:

>The entity returned with this response SHOULD include an indication of the request's current status

So the entity is not meant to be the actual result, it's meant to be a representation of the status of the request. So even if it were to contain either `true` or `false`, it would be wrong to interpret that as the actual result. (It would be odd for a 202 to return `true` or `false` but if it were to do so, that should, according to the HTTP spec, tell you something about the status. E.g., `true`, might mean 'done', and in that case it might be that when you fetch the completed result, it's actually `false`.)

So yes, I think we should fail for 202. I don't know what we should do for any of the various 3xx redirects. The problem with absolutely requiring a 200 is that we might fall foul of HTTP infrastructure—there are some systems that presume clients will handle a redirect. I _think_ it might be reasonable for us to say that because this is all web API stuff, and these are not endpoints for browser consumption, that we do not follow redirects. But this should probably be documented somewhere.



(Is the existing language that talks about throwing an exception correct? In the grand scheme of things we want the instance to become faulted—this may be triggered internally by throwing an exception that is then caught somewhere else inside of workflow, but is it meaningful to talk about throwing exceptions at this level of conceptual documentation, when workflow is conceived as a service?)

Reply:
Raise #111 to cover this.

Reply:
And also tweaked the wording to be clearer that throwing Exception => workflow instance is Faulted.